### PR TITLE
GEODE-85: Refactoring memcached into a separate module

### PR DESCRIPTION
**Pull request by upthewaterspout**
Moving memcached into a separate module. This module is still on the
classpath of the server by default, but it is not longer part of
geode-core.

### CODE REVIEW

**Comment by upthewaterspout** [comment]
@jxblum

File: geode-memcached/build.gradle [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+apply from: "${project.projectDir}/../gradle/publish.gradle"
+
+dependencies {
+  compile(platform(project(':boms:geode-all-bom')))
+  compile(project(':geode-core'))
+  distributedTestCompile(project(':geode-dunit'))
+  testCompile('org.mockito:mockito-core')
+  testCompile(project(':geode-junit'));
+  integrationTestCompile('net.spy:spymemcached')
+  integrationTestCompile(project(':geode-junit'));
```
Comment (PurelyApplied) [inline geode-memcached/build.gradle]: Minor, but I'd prefer these generally ordered as:

```
dependencies {
  compile(platform(project(':boms:geode-all-bom')))
  compile(project(':geode-core'))

  testCompile(project(':geode-junit'))
  testCompile('org.mockito:mockito-core')

  integrationTestCompile(project(':geode-junit'))
  integrationTestCompile('net.spy:spymemcached')

  distributedTestCompile(project(':geode-dunit'))
}
```
to be (semi-)consistent with the other dependency blocks throughout..

Also, the semicolons aren't necessary.

File: geode-core/src/main/resources/org/apache/geode/internal/sanctioned-geode-core-serializables.txt [hunk 2]
```diff
@@ -387,59 +387,6 @@ org/apache/geode/internal/datasource/TranxPoolCacheImpl,true,3295652525163658888
...
-org/apache/geode/internal/memcached/Reply$8,false
-org/apache/geode/internal/memcached/Reply$9,false
-org/apache/geode/internal/memcached/ResponseStatus,false
-org/apache/geode/internal/memcached/ResponseStatus$1,false
-org/apache/geode/internal/memcached/ResponseStatus$2,false
-org/apache/geode/internal/memcached/ResponseStatus$3,false
-org/apache/geode/internal/memcached/ResponseStatus$4,false
-org/apache/geode/internal/memcached/ResponseStatus$5,false
-org/apache/geode/internal/memcached/ResponseStatus$6,false
-org/apache/geode/internal/memcached/commands/ClientError,true,-2426928000696680541
```
Comment (PurelyApplied) [inline geode-core/src/main/resources/org/apache/geode/internal/sanctioned-geode-core-serializables.txt]: I really don't understand how the sanctioned serializables have historically worked, but is this an indicator that they might be expected to show up on the wire still?  Is this potentially an issue for rolling upgrades?  That was vaguely my understanding of what these tests were supposed to warn against.

Comment (upthewaterspout) [reply geode-core/src/main/resources/org/apache/geode/internal/sanctioned-geode-core-serializables.txt]: Good point - I'm not clear if these are sent over the wire. But I moved these sanctioned things into the geode-memcached project and added a AnalyzeSerializables* test to that project was well to validate them.

File: geode-core/src/main/java/org/apache/geode/distributed/internal/AbstractDistributionConfig.java [hunk 3]
```diff
@@ -641,8 +640,8 @@ protected int checkSecurityLogLevel(int value) {
   @ConfigAttributeChecker(name = MEMCACHED_PROTOCOL)
   protected String checkMemcachedProtocol(String protocol) {
     if (protocol == null
-        || !protocol.equalsIgnoreCase(GemFireMemcachedServer.Protocol.ASCII.name())
-            && !protocol.equalsIgnoreCase(GemFireMemcachedServer.Protocol.BINARY.name())) {
+        || !protocol.equalsIgnoreCase("ASCII")
```
Comment (rhoughton-pivot) [inline geode-core/src/main/java/org/apache/geode/distributed/internal/AbstractDistributionConfig.java]: Here we are changing from a type to a string value. In `DistributionConfig.java` below, we define a `DEFAULT_MEMCACHED_PROTOCOL` as that static string value. Should we be using a static String or Enum to hold this across all files?

File: geode-memcached/build.gradle [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+apply from: "${project.projectDir}/../gradle/publish.gradle"
+
+dependencies {
+  api(platform(project(':boms:geode-all-bom')))
+  implementation(project(':geode-core'))
+  implementation('org.apache.logging.log4j:log4j-api')
+  implementation('com.github.stephenc.findbugs:findbugs-annotations')
+
+  testCompile('org.mockito:mockito-core')
```
Comment (rhoughton-pivot) [inline geode-memcached/build.gradle]: We're so good, using `implementation` here. Can we use `testImplementation`, `integrationTestImplementation`, etc, as well?

File: geode-core/src/main/java/org/apache/geode/distributed/internal/AbstractDistributionConfig.java (hunk 3 above)
Comment (upthewaterspout) [reply geode-core/src/main/java/org/apache/geode/distributed/internal/AbstractDistributionConfig.java]: The enum is part of the geode-memcached subproject now, so we can't reference it in geode-core. It's also unfortunately public API and a subclass of GeodeMemcahedServer, so we can't reasonably leave it in core either.

File: geode-memcached/build.gradle (hunk 4 above)
Comment (upthewaterspout) [reply geode-memcached/build.gradle]: Fixed!

**Pull request merged**